packages/react-form-renderer/src/form-renderer/form-renderer.js

```diff
@@ -1,77 +1,132 @@
   const registeredFields = useRef({});
   const focusDecorator = useRef(createFocusDecorator());
-  let schemaError;
+  const validatorMapperMerged = useMemo(() => ({
+    ...defaultValidatorMapper,
+    ...validatorMapper
+  }), [validatorMapper]);
+  const mutatorsMerged = useMemo(() => ({
+    ...arrayMutators,
+    ...mutators
+  }), [mutators]);
+  const decoratorsMerged = useMemo(() => ([
+    focusDecorator.current,
+    ...(Array.isArray(decorators) ? decorators : [])
+  ]), [decorators]);
+
+  const handleSubmitCallback = useCallback((values, formApi, ...args) => {
+    return !isFunc(onSubmit) ? undefined : onSubmit(values, {...formApi, fileInputs}, ...args);
+  }, [onSubmit, fileInputs]);
+
+  const handleCancelCallback = useCallback((getState) => (...args) => {
+    return !isFunc(onCancel) ? void 0 : onCancel(getState().values, ...args);
+  }, [onCancel]);
+
+  const handleResetCallback = useCallback((reset) => (...args) => {
+    reset();
+    return !isFunc(onReset) ? void 0 : onReset(...args);
+  }, [onReset]);
 
-  const setRegisteredFields = (fn) => (registeredFields.current = fn({ ...registeredFields.current }));
-  const internalRegisterField = (name) => {
-    setRegisteredFields((prev) => (prev[name] ? { ...prev, [name]: prev[name] + 1 } : { ...prev, [name]: 1 }));
-  };
+  const handleErrorCallback = useCallback((...args) => {
+    // eslint-disable-next-line no-console
+    console.error(...args);
+    return !isFunc(onError) ? void 0 : onError(...args);
+  }, [onError]);
+
+  const registerInputFile = useCallback((name) => {
+    setFileInputs((prevFiles) => [...prevFiles, name]);
+  }, []);
+
+  const unRegisterInputFile = useCallback((name) => {
+    setFileInputs((prevFiles) => [
+      ...prevFiles.splice(prevFiles.indexOf(name))
+    ]);
+  }, []);
 
-  const internalUnRegisterField = (name) => {
-    setRegisteredFields(({ [name]: currentField, ...prev }) => (currentField && currentField > 1 ? { [name]: currentField - 1, ...prev } : prev));
-  };
+  const setRegisteredFields = useCallback((fn) => {
+    return registeredFields.current = fn({...registeredFields.current});
+  }, []);
 
-  const internalGetRegisteredFields = () =>
-    Object.entries(registeredFields.current).reduce((acc, [name, value]) => (value > 0 ? [...acc, name] : acc), []);
+  const internalRegisterField = useCallback((name) => {
+    setRegisteredFields((prev) => (
+      prev[name] ? {...prev, [name]: prev[name] + 1} : {...prev, [name]: 1})
+    );
+  }, []);
 
-  const validatorMapperMerged = { ...defaultValidatorMapper, ...validatorMapper };
+  const internalUnRegisterField = useCallback((name) => {
+    setRegisteredFields(({[name]: currentField, ...prev}) => (
+      currentField && currentField > 1 ? {[name]: currentField - 1, ...prev} : prev
+    ));
+  }, []);
+
+  const internalGetRegisteredFields = useCallback(() => {
+    const fields = registeredFields.current;
+    Object.entries(fields).reduce((acc, [name, value]) => (
```

Comment:
@zgover  This function is missing `return` that is why the submit tests in wizard are failing.